### PR TITLE
Create subproject relationship via APIv3 endpoint

### LINKED ISSUE #6157: api v3 is missing add a subproject
## Details

Adding a project as a Subproject of another project is still a manual operation.

From the docs:
In the admin dashboard for your project, select “Subprojects” from the menu. From this page you can add a subproject by typing in the project slug.

The API v3 Only has the method for: "Subprojects listing"

## Expected Result

In order to take advantage of subprojects we will need an API method, so that new projects that are added can be properly placed under their parent project

## Actual Result

Currently an admin would need to monitor new projects as they add their first docs and add them as a subproject by hand.

### PR DESCRIPTION
**Pull request by humitos**
The main goal of this PR is to add the ability to create a subproject relationship between two `Project` objects.

This PR is based on #5952 (https://github.com/readthedocs/readthedocs.org/tree/humitos/apiv3/project-update-endpoint) because it needs the "per object permissions check" changes that PR contains.

There are some things to note about this PR:

* `/api/v3/projects/pip/subprojects/` has changed its response. It was returning a list of `Project` but we found it was not 100% accurate and it was better to return a list of `ProjectRelationship` objects that includes the `alias` and the `child` (`Project`) rendered itself.
* a subproject relantionship can be create _only_ if the user is maintainer of both projects (parent and child)
* a subproject relationship can be deleted _only_ if the user is maintainer of the parent project

As a side note, while working on the tests for this PR, I found a but in the `RelatedProjectQuerySetBase` (see commit https://github.com/readthedocs/readthedocs.org/commit/0d42c1b9c17a6e5bd7ad80fac8d3d6ea353a61ee).

Closes #6157

### CODE REVIEW

**Comment by dwt2** [comment]
Would it also make sense to prevent project deletion if a project has one or more subprojects? I'm not sure if I think having to delete each subproject first would be terribly cumbersome or not, but it may be a good safeguard to have.

**Comment by stsewd** [comment]
@dwt2 currently if a project (main project) has submodules, and it is deleted, all its subprojects become normal projects.

**Comment by dwt2** [comment]
> @dwt2 currently if a project (main project) has submodules, and it is deleted, all its subprojects become normal projects.

Hmm, I think I would be easier on users to have the current default as an option, but also the option to pass a param to delete all subprojects with the top-level project too.

Let's say I intentionally want to delete a project and its 100 subprojects. With the current default I then have to do 100 more API operations to delete each (new) top-level project. Doable, but tedious.

**Comment by humitos** [comment]
> I think you are missing some more checks

I added these checks and tests for them as well. Please, review the PR when you have some time.

> Also, do we allow to add translations? There are some similar checks to be done too.

Currently, no. We are only listing translations under `/projects/<slug>/translations/`

**Comment by stsewd** [comment]
@humitos can you resolve the conflicts before review?

**Comment by humitos** [comment]
@stsewd done!

File: readthedocs/api/v3/serializers.py [hunk 1]
```diff
@@ -547,6 +547,127 @@ def get_subproject_of(self, obj):
...
+            )
+
+        # Check the parent project is not a subproject already
+        if self.parent_project.superprojects.exists():
+            raise serializers.ValidationError(
+                'Subproject nesting is not supported',
+            )
+        return data
+
+    class Meta:
```
Comment (stsewd) [inline readthedocs/api/v3/serializers.py]: The meta class should be at the top

File: readthedocs/projects/querysets.py [hunk 2]
```diff
@@ -137,7 +137,8 @@ def _add_user_repos(self, queryset, user=None):
             return self.all()
         if user.is_authenticated:
             projects_pk = user.projects.all().values_list('pk', flat=True)
-            user_queryset = self.filter(project__in=projects_pk)
+            kwargs = {'%s__in' % self.project_field: projects_pk}
```
Comment (stsewd) [inline readthedocs/projects/querysets.py]:
```suggestion
            kwargs = {f'{self.project_field}__in': projects_pk}
```

File: docs/api/v3.rst [hunk 3]
```diff
@@ -583,15 +611,63 @@ Subprojects listing
...
+            "child": "subproject-child-slug",
+            "alias": "subproject-alias"
+        }
+
+    **Example response**:
+
+    `See Subproject details <#subproject-details>`_
+
+    :>json string child: slug of the child project in the relationship.
+    :>json string alias: optional slug alias to be used in the URL.
```
Comment (stsewd) [inline docs/api/v3.rst]: We should note that if the alias isn't given, we use the slug as alias.

File: readthedocs/api/v3/views.py [hunk 4]
```diff
@@ -244,11 +248,40 @@ class SubprojectRelationshipViewSet(APIv3Settings, NestedViewSetMixin,
...
+    queryset = ProjectRelationship.objects.all()
+
+    def get_serializer_class(self):
+        """
+        Return correct serializer depending on the action.
+
+        For GET it returns a serializer with many fields and on POST,
+        it return a serializer to validate just a few fields.
+        """
+        if self.action == 'create':
```
Comment (stsewd) [inline readthedocs/api/v3/views.py]: aren't these constants somewhere?

Comment (humitos) [reply readthedocs/api/v3/views.py]: I've never seen them used as constants. I did a quick grep on their source code and I found that they use just the strings as well :'( : https://github.com/encode/django-rest-framework/blob/e57c1505fc4ed957332ae547a35c8713acfdf30c/rest_framework/schemas/coreapi.py#L37

**Comment by stsewd** [comment]
hmm, why wasn't this merged to master but `apiv3/project-update-endpoint`?

**Pull request merged**